src/Player.ts

```diff
@@ -75,68 +75,69 @@ class Player extends EventEmitter<PlayerEvents> {
      */
     private _handleVoiceState(oldState: VoiceState, newState: VoiceState): void {
         const queue = this.getQueue(oldState.guild.id);
-        if (!queue) return;
+        if (!queue || !queue.connection) return;
+        
+        if (oldState.channelId && !newState.channelId && newState.member.id === newState.guild.me.id) {
+            queue.destroy();
+            return void this.emit("botDisconnect", queue);
+        }
+        
+        if (!oldState.channelId && newState.channelId && newState.member.id === newState.guild.me.id) {
+            if (newState.serverMute || !newState.serverMute) {
+                queue.setPaused(newState.serverMute);
+            } else if (newState.suppress || !newState.suppress) {
+                if (newState.suppress) newState.guild.me.voice.setRequestToSpeak(true).catch(Util.noop);
+                queue.setPaused(newState.suppress);
+            }
+        }
 
-        if (oldState.channelId && newState.channelId && oldState.channelId !== newState.channelId) {
+        if (oldState.channelId === newState.channelId && newState.member.id === newState.guild.me.id) {
+            if (oldState.serverMute !== newState.serverMute) {
+                queue.setPaused(newState.serverMute);
+            }
+            else if (oldState.suppress !== newState.suppress) {
+                if (newState.suppress)
+                    newState.guild.me.voice.setRequestToSpeak(true).catch(Util_1.Util.noop);
+                queue.setPaused(newState.suppress);
+            }
+        }
+        
+        if (!newState.channelId && oldState.channelId === queue.connection.channel.id) {
+            if (!Util_1.Util.isVoiceEmpty(queue.connection.channel)) return;
+            const timeout = setTimeout(() => {
+                if (!Util_1.Util.isVoiceEmpty(queue.connection.channel)) return;
```

Comment:
This wont compile since it isnt valid in typescript. Please double check/run build command before making a pull request